Zend/zend_compile.c

```diff
@@ -1204,15 +1209,25 @@ zend_string *zend_type_to_string_resolved(zend_type type, zend_class_entry *scop
 	if (type_mask & MAY_BE_ARRAY) {
 		str = add_type_string(str, ZSTR_KNOWN(ZEND_STR_ARRAY));
 	}
+
+	if (type_mask == MAY_BE_SCALAR) {
```

Comment:
It's not easy to determine whether `scalar`/`number` or their union type counterpart should be displayed